rollup/system_contracts/l1_blocks.go

```diff
@@ -0,0 +1,144 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to query L1 chain ID, err = %w", err)
+	}
+	if got.Cmp(big.NewInt(0).SetUint64(l1ChainId)) != 0 {
+		return nil, fmt.Errorf("unexpected chain ID, expected = %v, got = %v", l1ChainId, got)
+	}
+
+	// get the L1Blocks ABI
+	l1BlocksAbi, err := abis.L1BlocksMetaData.GetAbi()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load L1Blocks ABI, err: %w", err)
+	}
+
+	worker := L1BlocksWorker{
+		ctx,
+		l1Client,
+		l1BlocksAbi,
+		confirmations,
+	}
+	return &worker, nil
+}
+
+func (w *L1BlocksWorker) GetLatestL1BlockNumberOnL2(state StateDB) uint64 {
+	return state.GetState(rcfg.L1BlocksAddress, rcfg.LatestBlockNumberSlot).Big().Uint64()
+}
+
+func (w *L1BlocksWorker) GetLatestConfirmedL1BlockNumber() (uint64, error) {
+	// confirmation based on "safe" or "finalized" block tag
+	if w.confirmations == rpc.SafeBlockNumber || w.confirmations == rpc.FinalizedBlockNumber {
+		tag := big.NewInt(int64(w.confirmations))
+		header, err := w.l1Client.HeaderByNumber(w.ctx, tag)
+		if err != nil {
+			return 0, err
+		}
+		if !header.Number.IsInt64() {
+			return 0, fmt.Errorf("received unexpected block number in BridgeClient: %v", header.Number)
+		}
+		return header.Number.Uint64(), nil
+	}
+
+	// confirmation based on latest block number
+	if w.confirmations == rpc.LatestBlockNumber {
+		number, err := w.l1Client.BlockNumber(w.ctx)
+		if err != nil {
+			return 0, err
+		}
+		return number, nil
+	}
+
+	// confirmation based on a certain number of blocks
+	if w.confirmations.Int64() >= 0 {
+		number, err := w.l1Client.BlockNumber(w.ctx)
+		if err != nil {
+			return 0, err
+		}
+		confirmations := uint64(w.confirmations.Int64())
+		if number >= confirmations {
+			return number - confirmations, nil
+		}
+		return 0, nil
+	}
+
+	return 0, fmt.Errorf("unknown confirmation type: %v", w.confirmations)
+}
+
+func (w *L1BlocksWorker) fetchL1BlockHeaderRlp(l1BlockNumber *big.Int) ([]byte, error) {
+	header, err := w.l1Client.HeaderByNumber(w.ctx, l1BlockNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get L1 block header, err: %w", err)
+	}
+	headerRlp, err := rlp.EncodeToBytes(header)
+	if err != nil {
+		return nil, fmt.Errorf("failed in RLP encoding of L1 block header, err: %w", err)
+	}
+	return headerRlp, nil
+}
+
+func (w *L1BlocksWorker) generateL1BlockMsg(l1BlockNumber uint64, nonce uint64) (*types.SystemTx, error) {
+	headerRlp, err := w.fetchL1BlockHeaderRlp(big.NewInt(int64(l1BlockNumber)))
+	if err != nil {
+		return nil, err
+	}
+	data, err := w.l1BlocksABI.Pack("setL1BlockHeader", headerRlp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to pack the calldata for setL1BlockHeader, err: %w", err)
+	}
+
+	return &types.SystemTx{
+		Sender: rcfg.SystemSenderAddress,
+		To:     rcfg.L1BlocksAddress,
+		Nonce:  nonce,
+		Data:   data,
+	}, nil
+}
+
+func (w *L1BlocksWorker) GenerateL1BlockMsgs(from, to uint64, state StateDB) ([]*types.SystemTx, error) {
```

Comment:
This method will introduce a lot of latency into block building, probably not suitable for production use. I suggest that `L1BlocksWorker` fetch blocks proactively (stored in memory) and returns finalized blocks from this in-memory buffer. We can change this later.

Reply:
moved it to sync service now